cstore_writer.c

```diff
@@ -310,104 +322,216 @@ CStoreWriteRow(TableWriteState *writeState, Datum *columnValues, bool *columnNul
 
-	CStoreWriteFooter(tempTableFooterFileName, writeState->tableFooter);
-
-	renameResult = rename(tempTableFooterFileName->data, tableFooterFilename->data);
-	if (renameResult != 0)
-	{
-		ereport(ERROR, (errcode_for_file_access(),
-						errmsg("could not rename file \"%s\" to \"%s\": %m",
-							   tempTableFooterFileName->data,
-							   tableFooterFilename->data)));
-	}
-
-	pfree(tempTableFooterFileName->data);
-	pfree(tempTableFooterFileName);
+	MemoryContextSwitchTo(oldContext);
 
 	MemoryContextDelete(writeState->stripeWriteContext);
 	list_free_deep(writeState->tableFooter->stripeMetadataList);
 	pfree(writeState->tableFooter);
-	pfree(writeState->tableFooterFilename->data);
-	pfree(writeState->tableFooterFilename);
 	pfree(writeState->comparisonFunctionArray);
 	FreeColumnBlockDataArray(writeState->blockDataArray, columnCount);
 	pfree(writeState);
 }
 
 
 /*
- * CStoreWriteFooter writes the given footer to given file. First, the function
- * serializes and writes the footer to the file. Then, the function serializes
- * and writes the postscript. Then, the function writes the postscript size as
- * the last byte of the file. Last, the function syncs and closes the footer file.
+ * CStoreWriteFooter writes the given footer data to relation footer file.
+ * First the function serializes the footer, the postscript, and the the
+ * postscript size as the last byte of the footer data.
+ * After preparing the footer data the function reads the current footer
+ * metadata to decide where to write to make sure that the current footer data
+ * is not overwritten. It writes the footer data to correct place and finally
+ * updates footer metadata about where footer data is stored. Reclamation of
+ * created StringInfo buffers are performed in caller function's memory context.
  */
 static void
-CStoreWriteFooter(StringInfo tableFooterFilename, TableFooter *tableFooter)
+CStoreWriteFooter(TableFooter *tableFooter, Relation relation, bool loggingEnabled)
 {
-	FILE *tableFooterFile = NULL;
 	StringInfo tableFooterBuffer = NULL;
 	StringInfo postscriptBuffer = NULL;
 	uint8 postscriptSize = 0;
-
-	tableFooterFile = AllocateFile(tableFooterFilename->data, PG_BINARY_W);
-	if (tableFooterFile == NULL)
-	{
-		ereport(ERROR, (errcode_for_file_access(),
-						errmsg("could not open file \"%s\" for writing: %m",
-							   tableFooterFilename->data)));
-	}
-
+	BlockNumber originalBlockCount = 0;
+	BlockNumber actualBlockCount = 0;
+	BlockNumber currentBlockNumber = 0;
+	StringInfo wholeFooter = makeStringInfo();
+	int32 dataLength = 0;
+	int32 dataOffset = 0;
+	const int32 blockDataSize =  CSTORE_PAGE_DATA_SIZE;
+	BlockNumber headerBlockNumber = InvalidBlockNumber;
+	Buffer headerBuffer = InvalidBuffer;
+	Page headerPage = NULL;
+	PageHeader headerBufferPageHeader = NULL;
+	char *headerPageContent = NULL;
+	uint32 startingBlock = InvalidBlockNumber;
+	uint32 blockCount = InvalidBlockNumber;
+	uint32 newStartingBlock = InvalidBlockNumber;
```

Comment:
Naming suggestion:
- `startingBlock` -> `previousBlockOffset` or `previousStartingBlock`
- `newStartingBlock` -> `blockOffset` or `startingBlock`

Or if you want to use `new.*`, update all related variable names to reflect the relationship. I am not against `new.*`, just make variable naming uniform and more descriptive.